### PR TITLE
[FLINK-24586][table] SQL functions should return STRING instead of VARCHAR(2000)

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

There are some SQL functions which currently return VARCHAR(2000). With more strict CAST behavior from FLINK-24413, this could become an issue.

The following functions return VARCHAR(2000) and should be changed to return STRING instead:

* JSON_QUERY
* JSON_OBJECT
* JSON_ARRAY

There are also some more functions which should be evaluated:

* CHR
* REVERSE
* SPLIT_INDEX
* PARSE_URL
* FROM_UNIXTIME
* DECODE

## Brief change log

  - *Updates the return type of the SQL Function `SqlJsonArrayFunction`, `SqlJsonObjectFunction` and `SqlJsonQueryFunction`.*
  - *Updates the return type of the SQL Function `CHR`, `REVERSE`, `SPLIT_INDEX`, `PARSE_URL`, `FROM_UNIXTIME` and `DECODE`.*

## Verifying this change

  - *Updates the `JsonFunctionsITCase` to verfiy the return type of the SQL function `JSON_QUERY`, `JSON_OBJECT` and `JSON_ARRAY` is STRING.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by SteNicholas** [comment]
@Airblader , could you please take a look at this pull request? Thanks.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/type/FlinkReturnTypes.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/type/FlinkReturnTypes.java:107]: Perhaps name it `VARCHAR_FORCE_NULLABLE` to make more explicit the nullability behaviour?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonQueryFunction.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonQueryFunction.java:37]: I wonder if you can just create a wrapper for `SqlFunction` to avoid copy-pasting `SqlJsonQueryFunction` and `SqlJsonArrayFunction`. If you check the code for `SqlFunction`, you see that the return type inference is used only for:

* `SqlReturnTypeInference getReturnTypeInference()` which is just a getter
* `RelDataType inferReturnType(SqlOperatorBinding opBinding)` which is a wrapper call for SqlReturnTypeInference.inferReturnType`.

WDYT?

**Review by slinkydeveloper — CHANGES_REQUESTED** [review]
Can you also test the return type now is correct for the string functions you changed the return type? `REVERSE`, `CHR`, `SPLIT_INDEX`, `PARSE_URL`, `FROM_UNIXTIME` and `DECODE`?

Also check in `FlinkSqlOperatorTable` for any usages of `ReturnTypes.cascade(ReturnTypes.explicit(SqlTypeName.VARCHAR), SqlTypeTransforms.FORCE_NULLABLE)` and replace them with the new constant

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/JsonFunctionsITCase.java
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/JsonFunctionsITCase.java:342]: The last argument can just be omitted if it's the same as the second last one.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/type/FlinkReturnTypes.java
Comment (SteNicholas) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/type/FlinkReturnTypes.java:107]: @slinkydeveloper , the `VARCHAR_FORCE_NULLABLE` makes sense to me.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonQueryFunction.java
Comment (SteNicholas) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonQueryFunction.java:37]: @slinkydeveloper , I agree with the point you mention above. I considered about the wrapper and I'm not sure whether creating a wrapper for `SqlFunction` has no problem. But I will try to create a wrapper for `SqlFunction`.

Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonQueryFunction.java:37]: Thank you! If we can fix it wrapping the class, it's a better solution. Copying classes should always be our last resort :smile:

**Review by slinkydeveloper — COMMENTED** [review]
This is looking good! I left some minor comments, you should also test the return types of REVERSE, CHR, SPLIT_INDEX, PARSE_URL, FROM_UNIXTIME and DECODE. You can add those tests in `StringFunctionsITCase`.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonObjectFunction.java
Comment (Airblader) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonObjectFunction.java:49]: If not done yet, please also add the TODOs for the Calcite upgrade into the corresponding JIRA issue directly; comments in the code will most likely not be seen.

**Comment by SteNicholas** [comment]
@slinkydeveloper , IMO, the previous tests include the return types of REVERSE, CHR, SPLIT_INDEX, PARSE_URL, FROM_UNIXTIME and DECODE because in this pull request I have updated the `testProcessingTimeInnerJoinWithOnClause.out` file. Does it need to add new tests for these functions?

**Comment by SteNicholas** [comment]
@slinkydeveloper @Airblader , thanks for the detailed review. I have addressed the above comments. But the tests for `CHR`, `REVERSE`, `SPLIT_INDEX`, `PARSE_URL`, `FROM_UNIXTIME` and `DECODE` couldn't add to `StringFunctionsITCase` which is tested for the `BuiltInFunctionDefinition` not `SqlFunction`.

**Review by slinkydeveloper — APPROVED** [review]
For me we're good to go, once the build is green we can merge it

**Comment by slinkydeveloper** [comment]
@flinkbot run azure

**Comment by SteNicholas** [comment]
@slinkydeveloper , the failed test has nothing to do with this change. Please check again.

**Review by hililiwei — APPROVED** [review]

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonObjectFunction.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonObjectFunction.java:57]: introduce a wrapper for this class already. we should not start patching classes copied from Calcite.

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks @SteNicholas. One last comment from my side. Otherwise good to merge in the next iteration.

**Comment by SteNicholas** [comment]
@twalthr , thanks for the detailed review. I have addressed the last comment above mentioned. Please help to review again.

**Review by twalthr — APPROVED** [review]
LGTM thanks for the update @SteNicholas.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java
Comment (tsreaper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:1141]: This function is not changed but it is in the list of FLINK-24586. Shall we also change this?

**Review by tsreaper — COMMENTED** [review]
Hi all! Sorry for revisiting this PR but I think we might miss `JSON_VALUE`.

**Pull request closed without merging**
